File: backend/src/main/java/br/com/calciolari/datahub/imports/domain/hints/FilenameHintsParser.java
package br.com.calciolari.datahub.imports.domain.hints;

import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Best-effort filename hint extractor. Never throws; ambiguous names yield empty hints.
 *
 * <p>Documented patterns (anchored within the basename, extension ignored for matching):
 * <ul>
 *   <li>{@code dd_MM-dd_MM} — incomplete period (no year), e.g. {@code 01_07-20_07}</li>
 *   <li>{@code dd/MM-dd/MM} — same with slashes</li>
 *   <li>{@code dd_MM} or {@code dd/MM} — single incomplete date when not part of a range</li>
 * </ul>
 *
 * Year is never inferred from the clock.
 */
public final class FilenameHintsParser {

	private static final Pattern PERIOD_UNDERSCORE = Pattern.compile(
			"(?<!\\d)(\\d{2})_(\\d{2})-(\\d{2})_(\\d{2})(?!\\d)");
	private static final Pattern PERIOD_SLASH = Pattern.compile(
			"(?<!\\d)(\\d{2})/(\\d{2})-(\\d{2})/(\\d{2})(?!\\d)");
	private static final Pattern SINGLE_UNDERSCORE = Pattern.compile("(?<!\\d)(\\d{2})_(\\d{2})(?!\\d)");
	private static final Pattern SINGLE_SLASH = Pattern.compile("(?<!\\d)(\\d{2})/(\\d{2})(?!\\d)");

	public FilenameHints parse(String originalFilename) {
		String preserved = originalFilename == null ? "" : originalFilename;
		String basename = stripDirectory(preserved);
		String stem = stripExtension(basename);

		Optional<IncompleteDateRange> period = matchPeriod(stem);
		if (period.isPresent()) {
			return new FilenameHints(preserved, period, Optional.empty());
		}

		Optional<IncompleteDate> single = matchSingle(stem);
		return new FilenameHints(preserved, Optional.empty(), single);
	}

	private static Optional<IncompleteDateRange> matchPeriod(String stem) {
		Matcher m = PERIOD_UNDERSCORE.matcher(stem);
		if (!m.find()) {
			m = PERIOD_SLASH.matcher(stem);
			if (!m.find()) {
				return Optional.empty();
			}
		}
		return toRange(m.group(1), m.group(2), m.group(3), m.group(4));
	}

	private static Optional<IncompleteDate> matchSingle(String stem) {
		Matcher m = SINGLE_UNDERSCORE.matcher(stem);
		if (!m.find()) {
			m = SINGLE_SLASH.matcher(stem);
			if (!m.find()) {
				return Optional.empty();
			}
		}
		return toDate(m.group(1), m.group(2));
	}

	private static Optional<IncompleteDateRange> toRange(String d1, String m1, String d2, String m2) {
		Optional<IncompleteDate> start = toDate(d1, m1);
		Optional<IncompleteDate> end = toDate(d2, m2);
		if (start.isEmpty() || end.isEmpty()) {
			return Optional.empty();
		}
		return Optional.of(new IncompleteDateRange(start.get(), end.get()));
	}

	private static Optional<IncompleteDate> toDate(String dayText, String monthText) {
		int day;
		int month;
		try {
			day = Integer.parseInt(dayText);
			month = Integer.parseInt(monthText);
		}
		catch (NumberFormatException ex) {
			return Optional.empty();
		}
		if (day < 1 || day > 31 || month < 1 || month > 12) {
			return Optional.empty();
		}
		return Optional.of(new IncompleteDate(day, month));
	}

	private static String stripDirectory(String name) {
		// Only strip Windows-style directories. Forward slashes are kept so
		// documented {@code dd/MM} filename hints remain matchable (browsers
		// already send basenames without Unix path prefixes).
		int slash = name.lastIndexOf('\\');
		return slash >= 0 ? name.substring(slash + 1) : name;
	}

	private static String stripExtension(String basename) {
		int dot = basename.lastIndexOf('.');
		if (dot <= 0) {
			return basename;
		}
		return basename.substring(0, dot);
	}
}

File: backend/src/Calciolari.DataHub/Imports/Domain/Hints/FilenameHintsParser.cs
using System.Text.RegularExpressions;

namespace Calciolari.DataHub.Imports.Domain.Hints;

/// <summary>
/// Best-effort filename hint extractor. Never throws; ambiguous names yield empty hints.
/// Year is never inferred from the clock.
/// </summary>
public sealed class FilenameHintsParser
{
    private static readonly Regex PeriodUnderscore = new(
        @"(?<!\d)(\d{2})_(\d{2})-(\d{2})_(\d{2})(?!\d)",
        RegexOptions.Compiled);

    private static readonly Regex PeriodSlash = new(
        @"(?<!\d)(\d{2})/(\d{2})-(\d{2})/(\d{2})(?!\d)",
        RegexOptions.Compiled);

    private static readonly Regex SingleUnderscore = new(
        @"(?<!\d)(\d{2})_(\d{2})(?!\d)",
        RegexOptions.Compiled);

    private static readonly Regex SingleSlash = new(
        @"(?<!\d)(\d{2})/(\d{2})(?!\d)",
        RegexOptions.Compiled);

    private static readonly Regex ProductCode = new(
        @"AUDITORIA\s+(\d+)\b",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    public FilenameHints Parse(string? originalFilename)
    {
        var preserved = originalFilename ?? string.Empty;
        var basename = StripDirectory(preserved);
        var stem = StripExtension(basename);
        var productCodeHint = MatchProductCode(stem);

        var period = MatchPeriod(stem);
        if (period is not null)
        {
            return new FilenameHints(preserved, period, null, productCodeHint);
        }

        var single = MatchSingle(stem);
        return new FilenameHints(preserved, null, single, productCodeHint);
    }

    private static string? MatchProductCode(string stem)
    {
        var match = ProductCode.Match(stem);
        return match.Success ? match.Groups[1].Value : null;
    }

    private static IncompleteDateRange? MatchPeriod(string stem)
    {
        var match = PeriodUnderscore.Match(stem);
        if (!match.Success)
        {
            match = PeriodSlash.Match(stem);
            if (!match.Success)
            {
                return null;
            }
        }

        return ToRange(match.Groups[1].Value, match.Groups[2].Value, match.Groups[3].Value, match.Groups[4].Value);
    }

    private static IncompleteDate? MatchSingle(string stem)
    {
        var match = SingleUnderscore.Match(stem);
        if (!match.Success)
        {
            match = SingleSlash.Match(stem);
            if (!match.Success)
            {
                return null;
            }
        }

        return ToDate(match.Groups[1].Value, match.Groups[2].Value);
    }

    internal static IncompleteDateRange? ToRange(string d1, string m1, string d2, string m2)
    {
        var start = ToDate(d1, m1);
        var end = ToDate(d2, m2);
        if (start is null || end is null)
        {
            return null;
        }

        return new IncompleteDateRange(start, end);
    }

    internal static IncompleteDate? ToDate(string dayText, string monthText)
    {
        if (!int.TryParse(dayText, out var day) || !int.TryParse(monthText, out var month))
        {
            return null;
        }

        if (day < 1 || day > 31 || month < 1 || month > 12)
        {
            return null;
        }

        return IncompleteDate.Create(day, month);
    }

    private static string StripDirectory(string name)
    {
        for (var i = name.Length - 1; i >= 0; i--)
        {
            var c = name[i];
            if (c is not '/' and not '\\')
            {
                continue;
            }

            var dateSlash = i > 0 && i + 1 < name.Length
                && char.IsDigit(name[i - 1])
                && char.IsDigit(name[i + 1]);
            if (dateSlash)
            {
                continue;
            }

            return name[(i + 1)..];
        }

        return name;
    }

    private static string StripExtension(string basename)
    {
        var dot = basename.LastIndexOf('.');
        if (dot <= 0)
        {
            return basename;
        }

        return basename[..dot];
    }
}
